### PR TITLE
Create leap.md

### LINKED ISSUE #524: [Cpp] Extract Concepts from v2 exercise: leap
Extract the list of concepts required to solve the v2 C++ exercise: `leap`. For each concept, describe why it is required to solve the exercise. Keep in mind that an exercise can have multiple approaches (you can find community solutions here: `https://exercism.io/tracks/cpp/exercises/leap/solutions`). The goal of this task is to help determine the total list of concepts that make up `leap`.

The list of concepts should be stored in the following document: `languages/cpp/reference/exercise-concepts/leap.md`. [Use this link to create a PR to add this document][pr-link].

[pr-link]: ../new/master?filename=languages/cpp/reference/exercise-concepts/leap.md

### PR DESCRIPTION
**Pull request by siebenschlaefer**
closes #524

### CODE REVIEW

File: languages/cpp/reference/exercise-concepts/leap.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+bool leap::is_leap_year(int year)
+{
+    return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
+}
+```
+
+
+## Concepts
+
+- .h file: the most common approach is to use a `.h` for the declarations
```
Comment (AeroStun) [inline languages/cpp/reference/exercise-concepts/leap.md]: We gotta discuss whether we promote .cpp/.hpp pairs or not

File: languages/cpp/reference/exercise-concepts/leap.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+- Some students use three local variables that tell if `year` is divisible by `4`, `100`, and `400`.
+- Some implement their solution as an `inline` function in the `.h` file.
+
+
+## Common mistakes
+
+- submitting the `.cpp` file without the `.h` file.
+- forgetting to enable and/or run all the tests
+- no include guards
+- defining the function in the header without declaring it as `inline`
```
Comment (AeroStun) [inline languages/cpp/reference/exercise-concepts/leap.md]: (or `constexpr`, which infers `inline`)

File: languages/cpp/reference/exercise-concepts/leap.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+- `if (condition) return true; else return false;` instead of `return condition;`
+- `condition? true: false` instead of `condition`
+- `!(year % 4 == 0)` instead of `year % 4 != 0
+- `year % 4 == 0 && year % 100 == 0 && year % 400 == 0` instead of `year % 400 == 0`
+- complicated nested `if` statements
```
Comment (AeroStun) [inline languages/cpp/reference/exercise-concepts/leap.md]: /ternaries

Comment (siebenschlaefer) [reply languages/cpp/reference/exercise-concepts/leap.md]: Although I *personally* find the logical operators much more readable, some people might prefer the conditional operator, e.g. our own example solution (https://github.com/exercism/cpp/blob/master/exercises/leap/example.h).

File: languages/cpp/reference/exercise-concepts/leap.md (hunk 1 above)
Comment (AeroStun) [reply languages/cpp/reference/exercise-concepts/leap.md]: Personally I make the strict separation between .c/.h and .cpp/.hpp, rationale being that it allows me to identify the language of a given header in shared langs repos.

Comment (siebenschlaefer) [reply languages/cpp/reference/exercise-concepts/leap.md]: Exercism [promotes ](https://teams.exercism.io/values) "fluency". IMHO that means a student should be able to handle both. But I think .h/.cpp files separate interface and implementation and its better to learn them first.

Comment (siebenschlaefer) [reply languages/cpp/reference/exercise-concepts/leap.md]: From the [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-file-suffix): "SF.1: Use a .cpp suffix for code files and .h for interface files if your project doesn’t already follow another convention".

**Pull request merged**